### ISSUE TITLE
Update copy_from_pqclean

### ISSUE THREAD

**Issue by baentsch** [open-quantum-safe/liboqs#880]
I'd support this proposal from Basil Hess and solicit further feedback to the below from @dstebila , @xvzcf : 

> Using kyber-upstream is now indeed a simple task thanks to what you already prepared.
> I really like the copy_from_pqclean mechanism. The META.yml definitions are obviously a pqclean-"invention", but they are a nice way to make an easy path for integrating also other implementations/libraries. Perhaps it could be renamed to copy_from_upstream.
> 
> I'm slightly updating the script with the following features. Please let me know if anything doesn't make sense here:
>
>- Adding support to checkout the upstream code from a git-repository: copy_from_pqclean.yml will contain a section "upstreams" with several git-repos (pqclean, pqcrystals-kyber, pqcrystals-dilithium, etc...), and the git-commit hash that will be used. This will replace the PQCLEAN_DIR env. variable. Instead, the git repo location will be used (this can be a github-url, a local path on the filesystem, etc.).
> - Adding a "verify" switch to the script: The goal is to be able to prove that the code used in liboqs origins from an upstream location with a specific commit. This will basically repeat the copy_from_pqclean process, hash of the code that is checked out, and compare it with the code in the liboqs src-directory. For someone using liboqs it should be useful to be able to reproduce and verify from where the crypto code origins.

**Comment by dstebila** [comment]
That all sounds good to me!

**Comment by xvzcf** [comment]
Same here.

**Comment by jschanck** [comment]
I like the "verify" idea. Does liboqs maintain any patches on upstream code? If so then verify needs to apply those patches before the comparison.

If there are not any patches on upstream, and we intend to have a "verify" option that does not apply patches, then I'll have to take a different approach to silencing false positives in #881.

**Comment by dstebila** [comment]
> I like the "verify" idea. Does liboqs maintain any patches on upstream code? If so then verify needs to apply those patches before the comparison.

For all the things imported via copy\_from\_pqclean, no, we do not add any patches.  (Things directly imported, like SIKE, Picnic, FrodoKEM, do make patches before being added.)

**Comment by bhess** [comment]
I also thought about adding the possibility to apply patches before the comparison, if necessary. But if there is an alternative this might be cleaner.

For now the updated copy_from_pqclean script is on this branch https://github.com/bhess/liboqs/tree/feat.copy_from_upstream

The script is renamed to `copy_from_upstream` to reflect that other upstream sources are possible.
For now pqcrystals-kyber and pqcrystals-dilithium are added as additional upstream sources to pqclean. They are defined in the `copy_from_upstream.yml` file like this (ideally the commits will be a tagged version):

```
upstreams:
  -
    name: pqclean
    git_url: https://github.com/PQClean/PQClean.git
    git_branch: master
    git_commit: 3d7d2024fa892bd7f00dca3fff4122175f4a26dc
    kem_meta_path: 'crypto_kem/{pqclean_scheme}/META.yml'
    sig_meta_path: 'crypto_sign/{pqclean_scheme}/META.yml'
    kem_scheme_path: 'crypto_kem/{pqclean_scheme}'
    sig_scheme_path: 'crypto_sign/{pqclean_scheme}'
  -
    name: pqcrystals-kyber
    git_url: https://github.com/pq-crystals/kyber.git
    git_branch: master
    git_commit: 8c8d7da74ee92808eebb6128ca209fd5851f48dc
    kem_meta_path: '{pretty_name_full}_META.yml'
    kem_scheme_path: '.'
  -
    name: pqcrystals-dilithium
    git_url: https://github.com/pq-crystals/dilithium.git
    git_branch: master
    git_commit: bda9689704262b53f0942efdda6340b922acb58b
    sig_meta_path: '{pretty_name_full}_META.yml'
    sig_scheme_path: '.'
...
```

copy_from_upstream has two options: "copy" and "verify":

```
$ ./copy_from_upstream.py --help
usage: copy_from_upstream.py [-h] [-v VERBOSITY] [-k] {copy,verify}

positional arguments:
  {copy,verify}

optional arguments:
  -h, --help            show this help message and exit
  -v VERBOSITY, --verbosity VERBOSITY
  -k, --keep_data       Keep temporary files
```

Copy checks out the git-repos and copies the upstream-sources to the src/ folder. The implementation folders in src/kem and src/sig are prefixed with the upstream names.

Verify creates a temporary folder and copies the upstream-sources there. It then does a diff of each implementation-folder with the corresponding ones in src/.

For example, if saber_avx2 differs from pqclean-upstream (just an example), it shows a follows:
```
$ ./copy_from_upstream.py verify
-----
Total schemes: 151 - 150 match with upstream, 1 differ
-----
Implementations that differ from upstream:
Name: pqclean_saber_avx2, expected upstream: https://github.com/PQClean/PQClean.git - commit: 3d7d2024fa892bd7f00dca3fff4122175f4a26dc
-----
```

I plan to open a PR for review later.

**Comment by baentsch** [comment]
Closed via #883
